File: mobile/lib/models/support_ticket.dart
/// Mirrors backend/app/schemas/support_ticket.py SupportTicketMessageOut.
class SupportTicketMessage {
  final String id;
  final String senderId;
  final String? senderName;
  final String body;
  final DateTime? createdAt;

  SupportTicketMessage({
    required this.id,
    required this.senderId,
    this.senderName,
    required this.body,
    this.createdAt,
  });

  factory SupportTicketMessage.fromJson(Map<String, dynamic> json) {
    return SupportTicketMessage(
      id: json['id'] as String,
      senderId: json['sender_id'] as String,
      senderName: json['sender_name'] as String?,
      body: json['body'] as String,
      createdAt: json['created_at'] != null ? DateTime.tryParse(json['created_at'] as String) : null,
    );
  }
}

/// Mirrors backend/app/schemas/support_ticket.py SupportTicketOut.
class SupportTicket {
  final String id;
  final String raisedById;
  final String? raisedByName;
  final String subject;
  final String? orderId;
  final String? trackingNumber;
  final String status; // open | in_progress | resolved | closed
  final DateTime? resolvedAt;
  final DateTime? createdAt;
  final List<SupportTicketMessage> messages;

  SupportTicket({
    required this.id,
    required this.raisedById,
    this.raisedByName,
    required this.subject,
    this.orderId,
    this.trackingNumber,
    required this.status,
    this.resolvedAt,
    this.createdAt,
    this.messages = const [],
  });

  bool get isOpen => status == 'open' || status == 'in_progress';

  factory SupportTicket.fromJson(Map<String, dynamic> json) {
    return SupportTicket(
      id: json['id'] as String,
      raisedById: json['raised_by_id'] as String,
      raisedByName: json['raised_by_name'] as String?,
      subject: json['subject'] as String,
      orderId: json['order_id'] as String?,
      trackingNumber: json['tracking_number'] as String?,
      status: json['status'] as String? ?? 'open',
      resolvedAt: json['resolved_at'] != null ? DateTime.tryParse(json['resolved_at'] as String) : null,
      createdAt: json['created_at'] != null ? DateTime.tryParse(json['created_at'] as String) : null,
      messages: json['messages'] != null
          ? (json['messages'] as List).map((e) => SupportTicketMessage.fromJson(e as Map<String, dynamic>)).toList()
          : const [],
    );
  }
}

File: backend/app/models/support_ticket.py
import enum

from sqlalchemy import Column, String, Text, DateTime, ForeignKey, Enum
from sqlalchemy.orm import relationship

from app.database import Base
from app.models.base import TimestampMixin, UUID_TYPE, gen_uuid


class SupportTicketStatus(str, enum.Enum):
    open = "open"
    in_progress = "in_progress"
    resolved = "resolved"
    closed = "closed"


class SupportTicket(Base, TimestampMixin):
    __tablename__ = "support_tickets"

    id = Column(UUID_TYPE, primary_key=True, default=gen_uuid)

    raised_by_id = Column(UUID_TYPE, ForeignKey("users.id"), nullable=False, index=True)
    raised_by = relationship("User", foreign_keys=[raised_by_id])

    subject = Column(String(200), nullable=False)
    order_id = Column(UUID_TYPE, ForeignKey("orders.id", ondelete='CASCADE'), nullable=True)
    order = relationship("Order")

    status = Column(Enum(SupportTicketStatus, native_enum=False, length=20), default=SupportTicketStatus.open, index=True)
    resolved_at = Column(DateTime(timezone=True), nullable=True)

    messages = relationship(
        "SupportTicketMessage",
        back_populates="ticket",
        order_by="SupportTicketMessage.created_at",
        cascade="all, delete-orphan",
    )


class SupportTicketMessage(Base, TimestampMixin):
    __tablename__ = "support_ticket_messages"

    id = Column(UUID_TYPE, primary_key=True, default=gen_uuid)

    ticket_id = Column(UUID_TYPE, ForeignKey("support_tickets.id"), nullable=False, index=True)
    ticket = relationship("SupportTicket", back_populates="messages")

    sender_id = Column(UUID_TYPE, ForeignKey("users.id"), nullable=False)
    sender = relationship("User")

    body = Column(Text, nullable=False)
